File: packages/ts/src/messages/cosmos.ts
/**
 * Convenience message builders for the standard Cosmos SDK modules.
 *
 * These wrap `cosmjs-types` message types and return cosmjs
 * {@link EncodeObject}s with the correct on-chain `typeUrl`. The matching types
 * are already in cosmjs's {@link defaultRegistryTypes} (and therefore in
 * {@link qorechainRegistry}), so a tx built from these encodes without extra
 * registration.
 *
 * Builders are grouped by module and re-exported (see {@link messages.ts}) so
 * callers write e.g. `msg.staking.delegate({ ... })` or `msg.gov.vote({ ... })`.
 */

import type { EncodeObject } from "@cosmjs/proto-signing";

import { MsgSend, MsgMultiSend } from "cosmjs-types/cosmos/bank/v1beta1/tx";
import {
  MsgDelegate,
  MsgUndelegate,
  MsgBeginRedelegate,
} from "cosmjs-types/cosmos/staking/v1beta1/tx";
import {
  MsgWithdrawDelegatorReward,
  MsgSetWithdrawAddress,
  MsgFundCommunityPool,
} from "cosmjs-types/cosmos/distribution/v1beta1/tx";
import {
  MsgVote,
  MsgVoteWeighted,
  MsgDeposit,
  MsgSubmitProposal,
} from "cosmjs-types/cosmos/gov/v1/tx";
import {
  MsgGrant,
  MsgRevoke,
  MsgExec,
} from "cosmjs-types/cosmos/authz/v1beta1/tx";
import {
  MsgGrantAllowance,
  MsgRevokeAllowance,
} from "cosmjs-types/cosmos/feegrant/v1beta1/tx";
import { MsgTransfer } from "cosmjs-types/ibc/applications/transfer/v1/tx";

/** A ts-proto/cosmjs generated message exposing `fromPartial`. */
interface MsgType<T> {
  fromPartial(object: T): T;
}

/** Build a `{ typeUrl, value }` composer bound to a fixed typeUrl + message. */
function composer<T>(typeUrl: string, msg: MsgType<Partial<T>>) {
  return (value: Partial<T>): EncodeObject => ({
    typeUrl,
    value: msg.fromPartial(value),
  });
}

/** Bank module message builders. */
export const bank = {
  send: composer<MsgSend>("/cosmos.bank.v1beta1.MsgSend", MsgSend),
  multiSend: composer<MsgMultiSend>(
    "/cosmos.bank.v1beta1.MsgMultiSend",
    MsgMultiSend,
  ),
};

/** Staking module message builders. */
export const staking = {
  delegate: composer<MsgDelegate>(
    "/cosmos.staking.v1beta1.MsgDelegate",
    MsgDelegate,
  ),
  undelegate: composer<MsgUndelegate>(
    "/cosmos.staking.v1beta1.MsgUndelegate",
    MsgUndelegate,
  ),
  redelegate: composer<MsgBeginRedelegate>(
    "/cosmos.staking.v1beta1.MsgBeginRedelegate",
    MsgBeginRedelegate,
  ),
};

/** Distribution module message builders. */
export const distribution = {
  withdrawDelegatorReward: composer<MsgWithdrawDelegatorReward>(
    "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward",
    MsgWithdrawDelegatorReward,
  ),
  setWithdrawAddress: composer<MsgSetWithdrawAddress>(
    "/cosmos.distribution.v1beta1.MsgSetWithdrawAddress",
    MsgSetWithdrawAddress,
  ),
  fundCommunityPool: composer<MsgFundCommunityPool>(
    "/cosmos.distribution.v1beta1.MsgFundCommunityPool",
    MsgFundCommunityPool,
  ),
};

/** Governance (gov v1) module message builders. */
export const gov = {
  vote: composer<MsgVote>("/cosmos.gov.v1.MsgVote", MsgVote),
  voteWeighted: composer<MsgVoteWeighted>(
    "/cosmos.gov.v1.MsgVoteWeighted",
    MsgVoteWeighted,
  ),
  deposit: composer<MsgDeposit>("/cosmos.gov.v1.MsgDeposit", MsgDeposit),
  submitProposal: composer<MsgSubmitProposal>(
    "/cosmos.gov.v1.MsgSubmitProposal",
    MsgSubmitProposal,
  ),
};

/** Authz (authorization grants) module message builders. */
export const authz = {
  grant: composer<MsgGrant>("/cosmos.authz.v1beta1.MsgGrant", MsgGrant),
  revoke: composer<MsgRevoke>("/cosmos.authz.v1beta1.MsgRevoke", MsgRevoke),
  exec: composer<MsgExec>("/cosmos.authz.v1beta1.MsgExec", MsgExec),
};

/** Fee-grant module message builders. */
export const feegrant = {
  grant: composer<MsgGrantAllowance>(
    "/cosmos.feegrant.v1beta1.MsgGrantAllowance",
    MsgGrantAllowance,
  ),
  revoke: composer<MsgRevokeAllowance>(
    "/cosmos.feegrant.v1beta1.MsgRevokeAllowance",
    MsgRevokeAllowance,
  ),
};

/** IBC fungible-token-transfer message builders. */
export const ibc = {
  transfer: composer<MsgTransfer>(
    "/ibc.applications.transfer.v1.MsgTransfer",
    MsgTransfer,
  ),
};

File: packages/py/src/qorsdk/messages/cosmos.py
"""Convenience message composers for the standard Cosmos SDK modules.

These wrap the message types cosmpy bundles (``cosmpy.protos.cosmos.*`` and
``cosmpy.protos.ibc.*``) and return :class:`~qorechain.messages._composer.Msg`
objects with the correct on-chain ``type_url``. They mirror the TypeScript SDK's
``msg.staking.delegate(...)`` / ``msg.gov.vote(...)`` surface (snake_cased) so a
transaction built from either SDK carries identical messages.

Builders are grouped per module into a small namespace object.
"""

from __future__ import annotations

from types import SimpleNamespace

from cosmpy.protos.cosmos.authz.v1beta1.tx_pb2 import MsgExec, MsgGrant, MsgRevoke
from cosmpy.protos.cosmos.bank.v1beta1.tx_pb2 import MsgMultiSend, MsgSend
from cosmpy.protos.cosmos.distribution.v1beta1.tx_pb2 import (
    MsgFundCommunityPool,
    MsgSetWithdrawAddress,
    MsgWithdrawDelegatorReward,
)
from cosmpy.protos.cosmos.feegrant.v1beta1.tx_pb2 import (
    MsgGrantAllowance,
    MsgRevokeAllowance,
)
from cosmpy.protos.cosmos.gov.v1.tx_pb2 import (
    MsgDeposit,
    MsgSubmitProposal,
    MsgVote,
    MsgVoteWeighted,
)
from cosmpy.protos.cosmos.staking.v1beta1.tx_pb2 import (
    MsgBeginRedelegate,
    MsgDelegate,
    MsgUndelegate,
)
from cosmpy.protos.ibc.applications.transfer.v1.tx_pb2 import MsgTransfer

from ._composer import composer

#: Bank module message composers.
bank = SimpleNamespace(
    send=composer("/cosmos.bank.v1beta1.MsgSend", MsgSend),
    multi_send=composer("/cosmos.bank.v1beta1.MsgMultiSend", MsgMultiSend),
)

#: Staking module message composers.
staking = SimpleNamespace(
    delegate=composer("/cosmos.staking.v1beta1.MsgDelegate", MsgDelegate),
    undelegate=composer("/cosmos.staking.v1beta1.MsgUndelegate", MsgUndelegate),
    redelegate=composer(
        "/cosmos.staking.v1beta1.MsgBeginRedelegate", MsgBeginRedelegate
    ),
)

#: Distribution module message composers.
distribution = SimpleNamespace(
    withdraw_delegator_reward=composer(
        "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward",
        MsgWithdrawDelegatorReward,
    ),
    set_withdraw_address=composer(
        "/cosmos.distribution.v1beta1.MsgSetWithdrawAddress", MsgSetWithdrawAddress
    ),
    fund_community_pool=composer(
        "/cosmos.distribution.v1beta1.MsgFundCommunityPool", MsgFundCommunityPool
    ),
)

#: Governance (gov v1) module message composers.
gov = SimpleNamespace(
    vote=composer("/cosmos.gov.v1.MsgVote", MsgVote),
    vote_weighted=composer("/cosmos.gov.v1.MsgVoteWeighted", MsgVoteWeighted),
    deposit=composer("/cosmos.gov.v1.MsgDeposit", MsgDeposit),
    submit_proposal=composer("/cosmos.gov.v1.MsgSubmitProposal", MsgSubmitProposal),
)

#: Authz (authorization grants) module message composers.
authz = SimpleNamespace(
    grant=composer("/cosmos.authz.v1beta1.MsgGrant", MsgGrant),
    revoke=composer("/cosmos.authz.v1beta1.MsgRevoke", MsgRevoke),
    exec=composer("/cosmos.authz.v1beta1.MsgExec", MsgExec),
)

#: Fee-grant module message composers.
feegrant = SimpleNamespace(
    grant=composer("/cosmos.feegrant.v1beta1.MsgGrantAllowance", MsgGrantAllowance),
    revoke=composer("/cosmos.feegrant.v1beta1.MsgRevokeAllowance", MsgRevokeAllowance),
)

#: IBC fungible-token-transfer message composers.
ibc = SimpleNamespace(
    transfer=composer("/ibc.applications.transfer.v1.MsgTransfer", MsgTransfer),
)

__all__ = [
    "bank",
    "staking",
    "distribution",
    "gov",
    "authz",
    "feegrant",
    "ibc",
]
